nntrainer/layers/layer_context.cpp

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file   layer_context.cpp
+ * @date   26 July 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is the layer context for each layer
+ */
+
+#include <layer_context.h>
+#include <var_grad.h>
+#include <weight.h>
+
+namespace nntrainer {
+
+/**
+ * @brief Get the Weight tensor object
+ *
+ * @param idx Identifier of the weight
+ * @return Tensor& Reference to the weight tensor
+ */
+Tensor &RunLayerContext::getWeight(unsigned int idx) const {
+  return weights[idx]->getVariableRef();
+}
+
+/**
+ * @brief Get the Weight Gradient tensor object
+ *
+ * @param idx Identifier of the weight
+ * @return Tensor& Reference to the weight grad tensor
+ */
+Tensor &RunLayerContext::getWeightGrad(unsigned int idx) const {
+  if (!weights[idx]->hasGradient())
+    throw std::invalid_argument(
+      "Requesting gradient for a non-trainable weight.");
+  return weights[idx]->getGradientRef();
+}
+
+/**
+ * @brief Get regularization loss for the weight
+ *
+ * @param idx Identifier of the weight
+ * @return float Value of the loss
+ */
+float RunLayerContext::getWeightRegularizationLoss(unsigned int idx) const {
+  if (weights[idx]->hasGradient())
+    return weights[idx]->getRegularizationLoss();
+
+  return 0;
```

Comment:
I wonder if it is valid to call for regularizationloss when they don't have gradient?
If it is, I guess we don't need check but return `weights[idx]->getRegularizationLoss()`?

Reply:
A weight can be initialized to be non-trainable but with regularization for now. I guess if that is stopped, then what you said should work.

Reply:
Cool